src/Utility/Hash.php

```diff
@@ -1146,6 +1146,86 @@ public static function mergeDiff(array $data, array $compare): array
         return $data;
     }
 
+    /**
+     * Validates data types for each key in $data against $types.
+     *
+     * Every key in $data must have a corresponding key in $types
+     * and the value in $data must match the expected type.
+     *
+     * Keys with mismatched or missing types are returned.
+     *
+     * Types can be one of the following or an array of possible types:
+     *
+     *  - 'null'
+     *  - 'array'
+     *  - 'string'
+     *  - 'float'
+     *  - 'int'
+     *  - 'bool'
+     *
+     *  If not one of the above built-in types, performs an `instanceof` check against the type.
+     *
+     * @param array $data The data to validate
+     * @param array $types The types for each expected key
+     * @param bool $strict Require all keys in both $data and $types to match
+     * @return array
+     */
+    public static function validate(array $data, array $types, bool $strict = false): array
+    {
+        $errors = [];
+        foreach ($data as $key => $value) {
+            $keyTypes = (array)($types[$key] ?? []);
+            if ($strict && !$keyTypes) {
+                $errors[] = $key;
+                continue;
+            }
+
+            foreach ($keyTypes as $keyType) {
+                if (static::validateValue($keyType, $value)) {
+                    continue 2;
+                }
+            }
+            $errors[] = $key;
+        }
+
+        if ($strict) {
+            $errors = array_merge($errors, array_keys(array_diff($types, $data)));
+        }
+
+        return $errors;
+    }
+
+    /**
+     * Validates value against type
+     *
+     * @param string $type Expected type
+     * @param mixed $value Value to check
+     * @return bool
+     */
+    protected static function validateValue(string $type, $value): bool
+    {
+        if ($type === 'null') {
```

Comment:
A switch case would be nicer to read.

Reply:
I had limited choice with 7.4.

Reply:
Wait, sorry, I was thinking match since I originally wrote that. Yea, we can use a switch sine the type is always a string.